### PR TITLE
adding abcd data conversion script

### PR DESCRIPTION
**Pull request by Arshitha**
@ericearl I'm back to solving #12 for the ABCD dataset. I downloaded the release 4.0 shared data package (only tabular data) from NDA. I've tested the conversion from `.txt` files to `.tsv` files and the script itself is working fine except for one problem. 

When the data are read into a pandas data frame, the columns with any missing values is converted to float type irrespective of whether it's string or integer type. I'm currently writing a function that type casts columns based on data type mentioned in the dictionary. I've a couple of questions about formatting of the final TSV though: 
1. The values in .txt files as downloaded from NDA are all enclosed in double quotes and therefore, all values are of the string type. Would it be better for the values in final TSV file to be encoded with `quoting=csv.QUOTE_ALL` ? 
2. Or, would it be more user-friendly for a column, let's say of type integer, to remain as integers in the final TSV? If we want to do this, then we'd also have to encode missing values with an integer value. (e.g., -999) 

Thoughts?

### CODE REVIEW

**Comment by Arshitha** [comment]
I like option 1 coz it's easier to implement and it won't require us to modify any formatting or missing value conventions.

**Comment by ericearl** [comment]
@Arshitha Thanks for this! I'll review it now! As for the TSV formatting options, as a proposed convention I think this repository should hold to these principles in priority order (and I will add an issue to add these tabular data conversion principles to the top-level README.md):

- Priority 1 = Never change data. Never change data types. Never add new fields to the original data.
- Priority 2 = Cleanly change all field separators to tabs.
- Priority 3 = Make data easier to use in TSV format while preserving the original intent.

So for this example, I feel option 1 is unnecessary. Perhaps `csv.QUOTE_NONE` would make more sense? On option 2, see Priority 1 above and "never change data types" means integers MUST remain integers. Also, encoding missing values is altering the data, so I do not like option 2. How about an "option 3"?

**Option 3**: Turn all comma separators to tab separators and left- and right-strip quotations (`whatever.lstrip('"').rstrip('"')`), preserving them all as strings and changing nothing.

File: ABCD/data_convert.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import argparse
+from pathlib import Path
+
+import pandas as pd
+
+
+def get_args():
+    parser = argparse.ArgumentParser(formatter_class=argparse.RawTextHelpFormatter, description=__doc__)
```
Comment (ericearl) [inline ABCD/data_convert.py:14]: I don't see a `__doc__` function anywhere. Or does this double underscore method do something I'm not aware of?

File: ABCD/data_convert.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    parser.add_argument('-t', '--tabular-data', type=Path, action='store', dest='tab_data', metavar='TAB_DATA_DIR',
+                        help='Path to directory with tabular data files in \'.txt\' format.')
```
Comment (ericearl) [inline ABCD/data_convert.py]: The README.md should dictate users put their downloaded data into the ABCD/ subfolder, so the script should look for it there as-is with no need tom be flexible.

File: ABCD/data_convert.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    parser.add_argument('-d', '--dictionaries', type=Path, action='store', dest='dicts', metavar='DICTS_DIR',
+                        default=Path('.'), help="Path to directory with data dictionaries in \'.csv\' format.")
```
Comment (ericearl) [inline ABCD/data_convert.py]: The README.md needs a section that tells the user how to prepare the data before running this script. This should include (in this case) the user MUST download the data dictionaries and use dictionary.py first before this. Then the dictionaries will be under this `ABCD/` subfolder in a `phenotype/` subfolder.

File: ABCD/data_convert.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    parser.add_argument('-o', '--output-dir', type=Path, action='store', dest='outdir', metavar='OUTPUT_DIR',
+                        default=Path('.'), help="Destination directory for the \'.tsv\' formatted files.")
```
Comment (ericearl) [inline ABCD/data_convert.py]: The outdir should always be the `phenotype/` folder.

File: ABCD/data_convert.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    args = parser.parse_args()
+
+    return args.tab_data.resolve(), args.dicts.resolve(), args.outdir.resolve()
+
+
+def data_summary(data_with_no_dict):
+    print(f"A total of {len(data_with_no_dict)} data files don't have a corresponding data dictionary.")
+    print(data_with_no_dict)
```
Comment (ericearl) [inline ABCD/data_convert.py]: Yikes! Did this ever happen in testing?

File: ABCD/data_convert.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    data_files_no_dict = []
+
+    for d in data_short_names:
+        expected_dict = data_dict_dir.joinpath(d + '.csv')  # constructing dictionary filename
+        if expected_dict not in data_dicts:
+            data_files_no_dict.append(d + '.txt')  # constructing data filename
+        else:
+            types_dict = {}  # dict of fields and data types
+            dict_df = pd.read_csv(expected_dict)
+            tsv_fields = dict_df['ElementName'].to_list()
```
Comment (ericearl) [inline ABCD/data_convert.py]: Oh! Is this the NDA data structure definitions? GitHub username tag me on your response so I can elaborate here.

**Review by ericearl — CHANGES_REQUESTED** [review]
@Arshitha Nice work! This will be a great addition! By convention, we need README.md edits for to make clear to the end user how they get to the point of running this script AND how to run this script. The convention should also be to make the `data_convert.py` require no arguments, but instead error-check if the user did their work prior to running the script. So you'll notice in some of my comments to expect the user has put data into the subfolder and things like that.

So basically if you can write a README addition this will be practically complete. Thanks again!

File: ABCD/data_convert.py (hunk 5 above)
Comment (Arshitha) [reply ABCD/data_convert.py]: It did actually. This was the output: 
```bash
A total of 14 data files don't have a corresponding data dictionary.
['abcd_mrisdp30201.txt', 'abcd_smrip30201.txt', 'abcd_mrisdp20201.txt', 'dataset_collection.txt', 'abcd_smrip20201.txt', 'package_info.txt', 'abcd_smrip10201.txt', 'abcd_mrisdp10201.txt', 'yabcdcovid19questionnaire01.txt', 'abcd_dmdtifp201.txt', 'mriqcrp20301.txt', 'guid_pseudoguid.txt', 'mriqcrp10301.txt', 'abcd_isc01.txt']
```

I guess dataset_collection.txt, package_info.txt can be ignored but the rest don't seem to have dictionaries associated with them.

File: ABCD/data_convert.py (hunk 4 above)
Comment (Arshitha) [reply ABCD/data_convert.py]: Noted!

File: ABCD/data_convert.py (hunk 3 above)
Comment (Arshitha) [reply ABCD/data_convert.py]: They could run `data_convert.py` before `dictionary.py` but they'll need to download the data from NDA and dictionaries from GitHub/NDA. I'll remove the argparser function.

File: ABCD/data_convert.py (hunk 1 above)
Comment (Arshitha) [reply ABCD/data_convert.py:14]: You're right! It should include __doc__ section for this to work. I could copy-paste contents of the README into __doc__ function? That way users can access it as command line help if they wish

File: ABCD/data_convert.py (hunk 6 above)
Comment (Arshitha) [reply ABCD/data_convert.py]: @ericearl what did you want to elaborate on ?

File: ABCD/data_convert.py (hunk 1 above)
Comment (ericearl) [reply ABCD/data_convert.py:14]: I typically put just a short description of the script in the argparse `description`. It populates into the help text every time you run with `-h` or`--help`.

File: ABCD/data_convert.py (hunk 5 above)
Comment (ericearl) [reply ABCD/data_convert.py]: @Arshitha That's fishy and we need to get to the bottom of it, yeah.

File: ABCD/data_convert.py (hunk 6 above)
Comment (ericearl) [reply ABCD/data_convert.py]: It sounds from your above comment that this uses the NDA data structure definition files. So I guess as long as it's clear to the end user they need those in the right place and to make sure they're there before trying this it's probably all good.

File: ABCD/data_convert.py (hunk 5 above)
Comment (Arshitha) [reply ABCD/data_convert.py]: I'll add this as an issue.

File: ABCD/data_convert.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    help_text()
+
+    # hard-coding file paths based on README instructions of dir organization
+    tab_data_dir = Path('data')
+    data_dict_dir = Path('nda_data_structure_definitions')
+    outdir = Path('phenotype')
+
+    # extract nda short names into a list
+    data_files = list(tab_data_dir.glob('*.txt'))
+    data_short_names = [i.name.split('.txt')[0] for i in
```
Comment (ericearl) [inline ABCD/data_convert.py]: I think you might want ["stem"](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.stem) here?

```suggestion
    data_short_names = [i.stem for i in
```

File: ABCD/data_convert.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+                        data_files]  # assert that this is indeed a list of strings.
+    data_dicts = list(data_dict_dir.glob('*.csv'))
+    data_files_no_dict = []
+    dict_no_data_files = []
+    for dict in data_dicts:
+        dict_prefix = dict.name.split('.')[0]
```
Comment (ericearl) [inline ABCD/data_convert.py:46]: I think you might want ["stem"](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.stem) here as well?

```suggestion
        dict_prefix = dict.stem
```

Comment (ericearl) [inline ABCD/data_convert.py]: I think you might want ["stem"](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.stem) here as well? And it is bad practice to override Python's `dict`. Perhaps `data_dict`?

```suggestion
    for data_dict in data_dicts:
        dict_prefix = data_dict.stem
```

File: ABCD/data_convert.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+            data_files_no_dict.append(d + '.txt')  # constructing data filename
+
+        # creating a list of lists where each sub-list is a formatted row
+        rows = []
+        curr_f = open(tab_data_dir.joinpath(d + '.txt'), 'r')
+        for idx, line in enumerate(curr_f.readlines()):
+            if idx != 1:  # skip over row describing fields
+                line = ','.join(line.strip('\n').split('\t'))
+                new_line = [word.strip('\"') for word in line.split(',') if not word == ""]
+                rows.append(new_line)
```
Comment (ericearl) [inline ABCD/data_convert.py]: You could either do a `curr_f.close()` (I think that's it anyway) or a `with open() as curr_f: something = curr_f.readlines()` or the below maybe. Either way, I think you should make sure that file handle/stream closes which this should handle.

Also, I added into the suggestion below to not trust commas for the separator on line 62. You should keep those things as `\t` because commas may be present in places, but tabs could never be present except as delimiters.

Last thing in this suggestion is the line 63 correction. Rather than stripping on all `\"` we should only strip the right-most and left-most `\"` because sometimes (if I remember right) there are quotations in the blocks of text for whatever reason.

```suggestion
        with open(tab_data_dir.joinpath(d + '.txt'), 'r') as curr_f:
            for idx, line in enumerate(curr_f.readlines()):
                if idx != 1:  # skip over row describing fields
                    no_newline = line.rstrip('\n')
                    new_line = [word.lstrip('"').rstrip('"') for word in no_newline.split('\t') if not word == ""]
                    rows.append(new_line)
```

File: ABCD/data_convert.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        # writing the list of lists as a TSV file
+        with open(outdir.joinpath(d + '.tsv'), 'w') as new_f:
+            wr = csv.writer(new_f, delimiter='\t', quoting=csv.QUOTE_MINIMAL)
+            wr.writerows(rows)
+    data_summary(data_files_no_dict, dict_no_data_files)
```
Comment (ericearl) [inline ABCD/data_convert.py:71]: @Arshitha This is troubling. Can we look at this together so we can figure out what to ask the NDA Help Desk?

File: ABCD/data_convert.py (hunk 8 above)
Comment (Arshitha) [reply ABCD/data_convert.py]: ah! thanks for pointing that out. DataSpell didn't warn me when I used dict and I didn't remember. Agreed!

**Review by ericearl — APPROVED** [review]
@Arshitha Thanks for this awesome work! If you include the suggestions I made (or something like them) and we talk about the data/no dict and dict/no data problems, I think this can be approved without the edits to the README yet. We would need to include a PR with the README edits soon after though. You can follow the general way I'm doing it here:

https://github.com/ericearl/dataset-phenotypes/blob/feature/add-hcp-data-convert/HCP/README.md#steps-to-produce-this-studys-tabular-data

Oh, and, I remember we had a discussion about this, but I just saw this entry [in the standard](https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#tabular-files):

> Missing and non-applicable values MUST be coded as n/a

So we should insert `n/a` into blank TSV fields for BIDS-standardization in a future PR. Raise it as an issue for now maybe?

**Comment by Arshitha** [comment]
How about I make that change as part of this PR?

File: ABCD/data_convert.py (hunk 9 above)
Comment (Arshitha) [reply ABCD/data_convert.py]: @ericearl yep, sounds good. We could have a meeting to discuss this further.

**Comment by ericearl** [comment]
> How about I make that change as part of this PR?

Sorry, I lost track with the many small changes. Which change are you referring to? The `n/a` change?

**Comment by Arshitha** [comment]
Yep, I should've been clear. I meant the 'n/a' change.

**Comment by ericearl** [comment]
Yeah, if that's easy, feel free to add it to this PR!

**Review by ericearl — APPROVED** [review]
Awesome and great! Thanks! Merging.
